README.md
# FFT-accelerated Interpolation-based t-SNE (FIt-SNE)
## Introduction
t-Stochastic Neighborhood Embedding ([t-SNE](https://lvdmaaten.github.io/tsne/)) is a highly successful method for dimensionality reduction and visualization of high dimensional datasets.  A popular [implementation](https://github.com/lvdmaaten/bhtsne) of t-SNE uses the Barnes-Hut algorithm to approximate the gradient at each iteration of gradient descent. We modified this implementation as follows:

* Computation of the N-body Simulation: Instead of approximating the N-body simulation using Barnes-Hut, we interpolate onto an equispaced grid and use FFT to perform the convolution, dramatically reducing the time to compute the gradient at each iteration of gradient descent. See the [this](http://gauss.math.yale.edu/~gcl22/blog/numerics/low-rank/t-sne/2018/01/11/low-rank-kernels.html) post for some intuition on how it works.
* Computation of Input Similiarities: Instead of computing nearest neighbors using vantage-point trees, we approximate nearest neighbors using the [Annoy](https://github.com/spotify/annoy) library. The neighbor lookups are multithreaded to take advantage of machines with multiple cores. Using "near" neighbors as opposed to strictly "nearest" neighbors is faster, but also has a smoothing effect, which can be useful for embedding some datasets (see [Linderman et al. (2017)](https://arxiv.org/abs/1711.04712)). If subtle detail is required (e.g. in identifying small clusters), then use vantage-point trees (which is also multithreaded in this implementation). 
* Early exaggeration: In [Linderman and Steinerberger (2017)](https://arxiv.org/abs/1706.02582), we showed that appropriately choosing the early exaggeration coefficient can lead to improved embedding of swissrolls and other synthetic datasets.
* Late exaggeration: Increasing the exaggeration coefficient late in the optimization process (e.g. after 800 of 1000 iterations) can improve separation of the clusters.

Check out the [preprint](https://arxiv.org/abs/1712.09005) for more details and some benchmarks.


## Installation
Note that the code has been tested for OS X and Linux, but not for Windows.
The only prerequisite is [FFTW](http://www.fftw.org/), which can be downloaded and installed from the website, prior to installation of this R pacakge.
This fork will install an R package ONLY.... for matlab and python implementations please see the original repo.
```r
    devtools::install_github("JulianSpagnuolo/FIt-SNE")
```
Check the examples given in the function documentation for usage demonstrations.



## References
If you use our software, please cite the original authors (not my fork):

George C. Linderman, Manas Rachh, Jeremy G. Hoskins, Stefan Steinerberger, Yuval Kluger. (2017). Efficient Algorithms for t-distributed Stochastic Neighborhood Embedding. (2017) *arXiv:1712.09005* ([link](https://arxiv.org/abs/1712.09005))

This implementation is derived from the Barnes-Hut implementation:

Laurens van der Maaten (2014). Accelerating t-SNE using tree-based algorithms. Journal of Machine Learning Research, 15(1):3221–3245. ([link](https://dl.acm.org/citation.cfm?id=2627435.2697068))


NEWS.md
# fftRtsne 1.0.1

* Added #include <functional> to tsne.cpp
* altered documentation of fftRtsne function
* Updated NAMESPACE

# fftRtsne 1.0.0

* Updated version number for release.

# fftRtsne 0.1.0

* Added a `NEWS.md` file to track changes to the package.





R/tsnehm.R
#' @export
#' @title tSNE Heatmap
#' @description Computes t-SNE Heatmap (EXPERIMENTAL)
#' @param expression_matrix Full expression matrix genes (rows) vs. cells (columns). Rownames should be gene names
#' @param goi goi Genes of Interest
#' @param tsne_embedding 1D t-SNE embedding
#' @param cell_labels Labels for each cell. Typically these are the cluster assignments for each cell, as obtained by 
#'                    dbscan on the 2D t-SNE. This way, the columns of the heatmap will have a color assigned to each of them,
#'                    and they can be mapped to a corresponding location on the 2D t-SNE
#' @param enrich For every gene in goi, find enrich number of genes that are close to that gene in the distance induced by the 1D t-SNE
#' @param breaks Number of bins
#' 
#' @examples
#' 
#' # Get a 1-dimensional embedding
#' ff <- fftRtsne(X=u.iris[,1:4], dims=1, fft_not_bh=T, max_iter=1000, rand_seed=42)
#' 
#' # Transposition of the expression matrix is NOT necessary if you are working with single cell RNA seq data.... i.e. your cells are in the columns not rows.
#' tsnehm(expression_matrix = t(u.iris[,1:4]), goi=colnames(u.iris[,1:4]), tsne_embedding = ff, cell_labels = u.iris$Species)
#' 
#' @return plotly object.
#' 
#' @import pdist
#' @import plyr
#' @importFrom RColorBrewer brewer.pal
#' @import heatmaply
#'
#require(gplots)
#library(pdist)
#library(plyr)
#library(RColorBrewer)
#library(heatmaply)
tsnehm <- function(expression_matrix, goi, tsne_embedding, cell_labels, enrich=0, breaks=100){
  
  nonempty_cells <- colSums(expression_matrix) > 0;
  expression_matrix <- expression_matrix[,nonempty_cells]
  tsne_embedding <- tsne_embedding[nonempty_cells]
  cell_labels <- cell_labels[nonempty_cells]
  
  nonempty_genes <- rowSums(expression_matrix) > 0;
  expression_matrix <- expression_matrix[nonempty_genes,]
  
  notinrows <- !(goi%in% rownames(expression_matrix));
  if (sum(notinrows) > 0){
    print(sprintf("The following rows are not in the matrix: %s", paste(goi[notinrows])))
  }
  goi <- goi[!notinrows]
  if (enrich ==0 ) {
    goidf <- data.frame(x=tsne_embedding, t(expression_matrix[goi,]))
  }else{
    goidf <- data.frame(x=tsne_embedding, t(expression_matrix))
  }
  group <- split(goidf, cut(goidf$x,breaks = breaks))
  bin_counts <- ldply(lapply(group,function(x) t(as.data.frame(colSums((x[,-1]))))), data.frame)
  bin_counts_s  <- sweep(bin_counts[,-1],2,colSums(bin_counts[,-1]), '/')
  if (enrich >0 ) {
    print("Now enriching")
    pdisttest <- pdist(t(as.matrix(bin_counts_s)),indices.A = goi, indices.B=1:ncol(bin_counts_s))
    sortedpdist <- t(apply(as.matrix(pdisttest), 1, order,decreasing=FALSE))
    enriched_genes <- unique(as.vector(sortedpdist[,1:(enrich +1)]))
    bin_counts_s <- bin_counts_s[,enriched_genes]
    
  }
  
  #assign a label to each column based on which of the cell_labels is the most common
  dbscan_tsne <- data.frame(x=tsne_embedding, y=cell_labels)
  dbscan_group <- split(dbscan_tsne, cut(dbscan_tsne$x,breaks = breaks))
  Mode <- function(x) {
    ux <- unique(x)
    ux[which.max(tabulate(match(x, ux)))]
  }
  group_labels <- unlist(lapply(dbscan_group, function(x) Mode(x[,2])))
  rownames(bin_counts_s) <- sprintf("bin:%s, label: %d", rownames(bin_counts_s), group_labels)
  group_labels[is.na(group_labels)] <- NA
  if (enrich>0){
  gene_colors <- rep(0, length(enriched_genes));
  }else{
    gene_colors <-c();
  }
  gene_colors[1:length(goi)] <- 1
  
  my_palette <- colorRampPalette(c("white", "red"))(n = 1000) ## convert to colorblind friendly col pal JS
  row_color_palette <- colorRampPalette(c("white", "blue")) ## convert to colorblind friendly col pal JS
  col_color_palette <- colorRampPalette(brewer.pal(n = max(group_labels+1,na.rm=TRUE),"Spectral")) ## convert to colorblind friendly col pal JS
  #toplot <- t(as.matrix(bin_counts_s>0.05,2) + 0)
  toplot <- t(as.matrix(bin_counts_s,2))
  
  ## Maybe covert this to a pheatmap call.... JS
  heatmaply(toplot,col_side_colors = group_labels, row_side_colors=gene_colors, 
            row_side_palette=row_color_palette, showticklabels = c(FALSE,TRUE),  
                  col=my_palette,  dendrogram='row', titleX =FALSE, RowV=FALSE,
                 show_legend=FALSE,hide_colorbar = TRUE,fontsize_row = 5,margins = c(70,50,NA,0))
}

R/fast_tsne.R
fftRtsne <- function(X, 
		     dims=2, perplexity=30, theta=0.5,
		     check_duplicates=TRUE,
		     max_iter=1000,
		     fft_not_bh = TRUE,
		     ann_not_vptree = TRUE,
		     stop_lying_iter=250,
		     exaggeration_factor=12.0, no_momentum_during_exag=FALSE,
		     start_late_exag_iter=-1.0,late_exag_coeff=1.0,
		     n_trees=50, search_k = -1,rand_seed=-1,
		     nterms=3, intervals_per_integer=1, min_num_intervals=50, 
		     data_path=NULL, result_path=NULL,
		     fast_tsne_path=NULL, nthreads=NULL)
{
  #' @export
  #' @title ffRtsne
  #' @author George Linderman, Cristoph H, Julian Spagnuolo
  #' @description Wrapper for fast_tsne binary, can perform both Barnes Hutt tSNE or the fast Fourier transform implementation.
  #' @param X numeric matrix or dataframe containing data to be tsne-fied
  #' @param dims integer. number of output dimensions to reduce to. Default is 2.
  #' @param perplexity integer. Default is 30.
  #' @param theta numeric. Set to 0 for exact.  If non-zero, then will use either Barnes Hut or FIt-SNE based on fft_not_bh. If Barnes Hut is used, then this determines the accuracy of BH approximation. Default is 0.5
  #' @param check_duplicates logical. Whether to check X for duplicates. Only set to FALSE if you are absolutely sure your data contains no duplicate entries. Default is TRUE.
  #' @param max_iter integer. Maximum iterations to run tsne over the data.
  #' @param fft_not_bh logical. Whether to run the Fast Fourier Transform tSNE (TRUE), or if set to FALSE, will use the BH implementation (slower). Default is TRUE.
  #' @param ann_not_vptree logical. If TRUE will approximate nearest neighbors using [Annoy](https://github.com/spotify/annoy), if FALSE the original BH-tSNE method using vantage-point trees will be used. Use vantage-point trees if fine detail is required (which is also multi-threaded in this implementation, so also quite fast if dimension <100). Default is TRUE
  #' @param stop_lying_iter integer. Determines the iteration at which early exageration is switch off. Default is 250.
  #' @param exaggeration_factor numeric. Early exageration coefficient, altering can lead to improved embedding of swissrolls and other synthetic datasets. Default is 12.0
  #' @param no_momentum_during_exag logical. If FALSE, momentum and other optimizations will be used in the gradient decent phase. If TRUE, standard gradient descent will be used. Useful for testing large exaggeration coefficients. Default is FALSE.
  #' @param start_late_exag_iter numeric. Iteration at which to start late exageration, if set to -1, late exageration will not be used. Default is -1.0
  #' @param late_exag_coeff numeric. Increasing this parameter may improve separation of the clusters in the reduced space. Default is 1.0
  #' @param n_trees integer. Default is 50
  #' @param search_k numeric. Default is -1.
  #' @param rand_seed numeric. Set a seed number for reproducibility, if -1 the random seed will be set using the system time. Default is -1.
  #' @param nterms integer. The number of interpolation points per sub-interval when using FIt-SNE Default is 3.
  #' @param intervals_per_integer integer. See Details, must be >0. Default is 1
  #' @param min_num_intervals integer. See Details. Default is 50
  #' @param data_path character. File path of a data file. If NULL, X will be written to a temp file. Default is NULL
  #' @param result_path character. File path to write results to, if NULL results will be written to a temp file prior to being read into R. Default is NULL
  #' @param fast_tsne_path character. File path to the location of the compiled fast_tsne_binary. This version of the forked git repo will compile the fast_tsne binary to the package installation directory, if this param is NULL then the function will default to this location. You can change this behaviour if you have a custom install of the original repo. Default is NULL
  #' @param nthreads integer. Number of threads used by computeGaussianPerplexity, if NULL, will use detectCores()-1 to determine the number of threads to use.
  #' @details 
  #' ann_not_vptree
  #'  Using "near" neighbors as opposed to strictly "nearest" neighbors is faster, but also has a smoothing effect,
  #'  which can be useful for embedding some datasets (see Linderman et al. (2017)).
  #'  If subtle detail is required (e.g. in identifying small clusters), then use vantage-point trees
  #'  (which is also multithreaded in this implementation).
  #' 
  #' min_num_intervals & intervals_per_integer
  #' 
  #' Let:
  #'    maxloc = ceiling(max(max(X)))
  #'    and
  #'    minloc = floor(min(min(X)))
  #'    i.e. the data points are in a minloc^n.dims by maxloc^n.dims interval/square. The number os intervals in each dimension
  #'    is either min_num_intervals or ceiling(maxloc - minloc)/intervals_per_integer whichever is larger.
  #'    
  #' @return Matrix of the reduced tSNE dimensions, ncol = dims, nrow = nrow(X).
  #' @references George C. Linderman, Manas Rachh, Jeremy G. Hoskins, Stefan Steinerberger, Yuval Kluger. (2017). Efficient Algorithms for t-distributed Stochastic Neighborhood Embedding. (2017) arXiv:1712.09005
  #' 
  #' @examples 
  #' \dontrun{
  #' # import the iris data set and remove any duplicates 
  #' library(ggplot2)
  #' library(ggthemes)
  #' data(iris)
  #' 
  #' u.iris <- unique(iris)
  #' 
  #' # For a standard BH-tsne
  #' 
  #' bh <- fftRtsne(X=u.iris[,1:4], dims = 2, fft_not_bh = FALSE, max_iter = 1000, rand_seed=42)
  #' u.iris$bh.tsne1 <- bh[,1]
  #' u.iris$bh.tsne2 <- bh[,2]
  #' 
  #' fig <- ggplot(u.iris, aes(x=bh.tsne1, y=bh.tsne2, colour=Species))
  #' fig +geom_point() +scale_color_colorblind() +theme_bw() +theme(aspect.ratio=1)
  #' 
  #' # For the fast Fourier transform tSNE
  #' 
  #' ff <- fftRtsne(X=u.iris[,1:4], dims=2, fft_not_bh=TRUE, max_iter=1000, rand_seed=42)
  #' u.iris$ff.tsne1 <- ff[,1]
  #' u.iris$ff.tsne2 <- ff[,2]
  #' 
  #' fig <- ggplot(u.iris, aes(x=ff.tsne1, y=ff.tsne2, colour=Species))
  #' fig +geom_point() +scale_color_colorblind() +theme_bw() +theme(aspect.ratio=1)
  #' }
  #' 
  #' @importFrom utils file_test
  #' @importFrom parallel detectCores
	
  if (is.null(data_path)) {
		data_path <- tempfile(pattern='fftRtsne_data_', fileext='.dat')
	}
	if (is.null(result_path)) {
		result_path <- tempfile(pattern='fftRtsne_result_', fileext='.dat')
	}
	#if (is.null(fast_tsne_path)) { ## No longer necessary with default install location using makefile. JS
	#	fast_tsne_path <- system2('which', 'fast_tsne', stdout=TRUE)
	#}
  if(is.null(fast_tsne_path)) ### Added in to default to R package install location. JS.
  {
    fast_tsne_path <- paste(normalizePath(find.package("fftRtsne")), "fast_tsne", sep="/")
  }
	fast_tsne_path <- normalizePath(fast_tsne_path)
	if (!file_test('-x', fast_tsne_path)) {
		stop(fast_tsne_path, " does not exist or is not executable; check your fast_tsne_path parameter")
	}

	if(is.null(nthreads)) ## added in to make maximal use of detected cores, minus one for safety :) JS
	{
	  nthreads <- detectCores()-1
	}
	
	is.wholenumber <- function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) < tol

	if (!is.numeric(theta) || (theta<0.0) || (theta>1.0) ) { stop("Incorrect theta.")}
	if (nrow(X) - 1 < 3 * perplexity) { stop("Perplexity is too large.")}
	if (!is.matrix(X)) ## changed to convert to matrix if it is a dataframe, added verbosity if you tried to give it something else. JS
	  { 
	    if(is.data.frame(X))
	    {
	      X <- as.matrix(X)
	    }
	    else(stop("Not a data.frame or matrix"))
	  } 
	if (!(max_iter>0)) { stop("Incorrect number of iterations.")}
	if (!is.wholenumber(stop_lying_iter) || stop_lying_iter<0) { stop("stop_lying_iter should be a positive integer")}
	if (!is.numeric(exaggeration_factor)) { stop("exaggeration_factor should be numeric")}
	if (!is.wholenumber(dims) || dims<=0) { stop("Incorrect dimensionality.")}
	if (search_k == -1) { search_k = n_trees*perplexity*3 }

	if (fft_not_bh){
	  nbody_algo = 2;
	}else{
	  nbody_algo = 1;
	}
	
	if (ann_not_vptree){
	  knn_algo = 1;
	}else{
	  knn_algo = 2;
	}
	tX = c(t(X))

	f <- file(data_path, "wb")
	n = nrow(X);
	D = ncol(X);
	writeBin(as.integer(n), f,size= 4)
	writeBin( as.integer(D),f,size= 4)
	writeBin( as.numeric(0.5), f,size= 8) #theta
	writeBin( as.numeric(perplexity), f,size= 8) #theta
	writeBin( as.integer(dims), f,size=4) #theta
	writeBin( as.integer(max_iter),f,size=4)
	writeBin( as.integer(stop_lying_iter),f,size=4)
	writeBin( as.integer(-1),f,size=4) #K
	writeBin( as.numeric(-30.0), f,size=8) #sigma
	writeBin( as.integer(nbody_algo), f,size=4)  #not barnes hut
	writeBin( as.integer(knn_algo), f,size=4) 
	writeBin( as.numeric(exaggeration_factor), f,size=8) #compexag
	writeBin( as.integer(no_momentum_during_exag), f,size=4) 
	writeBin( as.integer(n_trees), f,size=4) 
	writeBin( as.integer(search_k), f,size=4) 
	writeBin( as.integer(start_late_exag_iter), f,size=4) 
	writeBin( as.numeric(late_exag_coeff), f,size=8) 
	
	writeBin( as.integer(nterms), f,size=4) 
	writeBin( as.numeric(intervals_per_integer), f,size=8) 
	writeBin( as.integer(min_num_intervals), f,size=4) 
	tX = c(t(X))
	writeBin( tX, f) 
	writeBin( as.integer(rand_seed), f,size=4) 
	close(f) 

	flag= system2(command=fast_tsne_path, args=c(data_path, result_path, nthreads));
	if (flag != 0) {
		stop('tsne call failed');
	}
	f <- file(result_path, "rb")
	initialError <- readBin(f, integer(), n=1, size=8);
	n <- readBin(f, integer(), n=1, size=4);
	d <- readBin(f, integer(), n=1,size=4);
	Y <- readBin(f, numeric(), n=n*d);
	Yout <- t(matrix(Y, nrow=d));
	close(f)
	file.remove(data_path)
	file.remove(result_path)
	Yout
}





